### PR TITLE
RONDB-288: Use Faster Base64 library

### PR DESCRIPTION
**Pull request by olapiv**

### CODE REVIEW

**Comment by olapiv** [comment]
Will probably add benchmark unit tests in a different PR.

File: storage/ndb/rest-server/data-access-rondb/src/db-operations/pk/common.cpp [hunk 1]
```diff
@@ -336,26 +337,34 @@ RS_Status SetOperationPKCol(const NdbDictionary::Column *col, NdbOperation *oper
...
     size_t col_len               = col->getLength();
 
-    size_t decoded_size = boost::beast::detail::base64::decoded_size(encoded_str_len);
-    if (unlikely(decoded_size > BINARY_MAX_SIZE_IN_BYTES_DECODED)) {
-      return RS_CLIENT_ERROR(std::string(ERROR_008) + " " +
-                             "Decoded data length is greater than column length. " +
-                             "Column: " + std::string(col->getName()) +
-                             " Length: " + std::to_string(col->getLength()));
+    // The buffer in out has been allocated by the caller and is at least 3/4 the size of the input.
+    const int maxEncodedSize = ((4 * BINARY_MAX_SIZE_IN_BYTES_DECODED) / 3);
```
Comment (smkniazi) [inline storage/ndb/rest-server/data-access-rondb/src/db-operations/pk/common.cpp]: BINARY_MAX_SIZE_IN_BYTES_DECODED -> BINARY_MAX_SIZE_IN_BYTES

File: storage/ndb/rest-server/data-access-rondb/src/db-operations/pk/common.cpp [hunk 2]
```diff
@@ -336,26 +337,34 @@ RS_Status SetOperationPKCol(const NdbDictionary::Column *col, NdbOperation *oper
...
     char pk[BINARY_MAX_SIZE_IN_BYTES_DECODED];
     memset(pk, 0, col->getLength());
 
-    std::pair<std::size_t, std::size_t> ret =
-        boost::beast::detail::base64::decode(pk, encoded_str, encoded_str_len);
-    // make sure everything was decoded. 1 or 2 bytes of padding is not included in ret.second
-    require(ret.second >= encoded_str_len - 2 && ret.second <= encoded_str_len);
-    if (unlikely(ret.first > col_len)) {
+    size_t outlen = 0;
+    int result = base64_decode(encoded_str, encoded_str_len, (char *)&pk[0], &outlen, 0);
```
Comment (smkniazi) [inline storage/ndb/rest-server/data-access-rondb/src/db-operations/pk/common.cpp]: reinterpret_cast<char *>(pk)

File: storage/ndb/rest-server/data-access-rondb/src/db-operations/pk/common.cpp [hunk 3]
```diff
@@ -336,26 +337,34 @@ RS_Status SetOperationPKCol(const NdbDictionary::Column *col, NdbOperation *oper
...
+
+    if (unlikely(result == 0)) {
```
Comment (smkniazi) [inline storage/ndb/rest-server/data-access-rondb/src/db-operations/pk/common.cpp:361]: why "0" is considered an error here? https://github.com/aklomp/base64/blob/e77bd70bdd860c52c561568cffb251d88bba064c/lib/lib.c#L163

File: storage/ndb/rest-server/data-access-rondb/src/db-operations/pk/common.cpp [hunk 4]
```diff
@@ -374,13 +383,13 @@ RS_Status SetOperationPKCol(const NdbDictionary::Column *col, NdbOperation *oper
...
 
-    size_t decoded_size = boost::beast::detail::base64::decoded_size(encoded_str_len);
-
-    if (unlikely(decoded_size > KEY_MAX_SIZE_IN_BYTES_DECODED)) {
-      return RS_CLIENT_ERROR(std::string(ERROR_008) + " " +
-                             "Decoded data length is greater than column length. " +
-                             "Column: " + std::string(col->getName()) +
-                             " Length: " + std::to_string(col->getLength()));
+    // The buffer in out has been allocated by the caller and is at least 3/4 the size of the input.
+    const int maxEncodedSize = ((4 * KEY_MAX_SIZE_IN_BYTES_DECODED) / 3);
```
Comment (smkniazi) [inline storage/ndb/rest-server/data-access-rondb/src/db-operations/pk/common.cpp]: KEY_MAX_SIZE_IN_BYTES_DECODED --- > KEY_MAX_SIZE_IN_BYTE

File: storage/ndb/rest-server/data-access-rondb/src/db-operations/pk/common.cpp [hunk 5]
```diff
@@ -389,36 +398,36 @@ RS_Status SetOperationPKCol(const NdbDictionary::Column *col, NdbOperation *oper
...
     }
 
+    size_t outlen = 0;
     // leave first 1-2 bytes free for saving length bytes
-    std::pair<std::size_t, std::size_t> ret =
-        boost::beast::detail::base64::decode(pk + additional_len, encoded_str, encoded_str_len);
-    
-    // Make sure everything was decoded. 1 or 2 bytes of padding which is not included in ret.second
-    if (unlikely((ret.second < encoded_str_len - 2) || (ret.second > encoded_str_len))) {
+    int result = base64_decode(encoded_str, encoded_str_len, (char *)(&pk[0] + additional_len), &outlen, 0);
```
Comment (smkniazi) [inline storage/ndb/rest-server/data-access-rondb/src/db-operations/pk/common.cpp]: reinterpret_cast<char *>(pk + additional_len)

File: storage/ndb/rest-server/data-access-rondb/src/db-operations/pk/common.cpp [hunk 6]
```diff
@@ -389,36 +398,36 @@ RS_Status SetOperationPKCol(const NdbDictionary::Column *col, NdbOperation *oper
...
+
+    if (unlikely(result == 0)) {
```
Comment (smkniazi) [inline storage/ndb/rest-server/data-access-rondb/src/db-operations/pk/common.cpp:419]: same as above, doesn't zero represent a successful operation

**Review by smkniazi — COMMENTED** [review]
I did not get why we have to throw an error here ```if (unlikely(result == 0))```

**Review by smkniazi — DISMISSED** [review]

File: storage/ndb/rest-server/data-access-rondb/src/db-operations/pk/common.cpp [hunk 7]
```diff
@@ -384,15 +384,15 @@ RS_Status SetOperationPKCol(const NdbDictionary::Column *col, NdbOperation *oper
...
     if (unlikely(encoded_str_len > maxEncodedSize)) {
         return RS_CLIENT_ERROR(std::string(ERROR_008) +
                         " Encoded data length is greater than 4/3 of maximum binary size." +
                         " Column: " + std::string(col->getName()) +
-                        " Maximum binary size: " + std::to_string(KEY_MAX_SIZE_IN_BYTES_DECODED));
+                        " Maximum binary size: " + std::to_string(KEY_MAX_SIZE_IN_BYTES));
     }
 
-    char pk[KEY_MAX_SIZE_IN_BYTES_DECODED];
+    char pk[KEY_MAX_SIZE_IN_BYTES];
```
Comment (smkniazi) [inline storage/ndb/rest-server/data-access-rondb/src/db-operations/pk/common.cpp:408]: this one (line 395) was not supposed to change. KEY_MAX_SIZE_IN_BYTES --> KEY_MAX_SIZE_IN_BYTES_DECODED

File: storage/ndb/rest-server/data-access-rondb/src/db-operations/pk/common.cpp [hunk 8]
```diff
@@ -338,15 +338,15 @@ RS_Status SetOperationPKCol(const NdbDictionary::Column *col, NdbOperation *oper
...
     if (unlikely(encoded_str_len > maxEncodedSize)) {
         return RS_CLIENT_ERROR(std::string(ERROR_008) + " " +
                         "Encoded data length is greater than 4/3 of maximum binary size." +
                         " Column: " + std::string(col->getName()) +
-                        " Maximum binary size: " + std::to_string(BINARY_MAX_SIZE_IN_BYTES_DECODED));
+                        " Maximum binary size: " + std::to_string(BINARY_MAX_SIZE_IN_BYTES));
     }
 
-    char pk[BINARY_MAX_SIZE_IN_BYTES_DECODED];
+    char pk[BINARY_MAX_SIZE_IN_BYTES];
```
Comment (smkniazi) [inline storage/ndb/rest-server/data-access-rondb/src/db-operations/pk/common.cpp:355]: line 349 was not supposed to change. BINARY_MAX_SIZE_IN_BYTES --> BINARY_MAX_SIZE_IN_BYTES_DECODED

**Review by smkniazi — CHANGES_REQUESTED** [review]

**Review by mronstro — APPROVED** [review]

**Review by smkniazi — APPROVED** [review]

**Pull request merged**